### ISSUE TITLE
Distributed Training workflows

### ISSUE THREAD

**Issue by shivance** [keras-team/keras-nlp#858]
Language Models are compute demanding, Training on single GPU with small batch size is time taking. Luckily Kaggle provides us 2 T4 GPUs. Do we have any guide that demonstrates training KerasNLP models in **distributed** fashion?

cc : @mattdangerw @chenmoneygithub @abheesht17

**Comment by mattdangerw** [comment]
Yeah, I think there are two major axis of distribution to cover.

 - Model parallelism, where the a single model cannot fit in GPU/TPU memory and is hosted across many accelerators. We are actually working on some a guide for this using https://www.tensorflow.org/guide/dtensor_overview soon!
 - Data parallelism, where a batch is split onto multiple machines/accelerators and the gradients combined. This would be great to have some guides/examples for, and is ready today with the `tf.distrbute` API.

I'm not sure what the right deliverable is, maybe a keras.io example on `Data parallel training with KerasNLP and tf.distribute`? But we would welcome some examples here.

Adding @sampathweb who is knowledgeable on the distribute side.

**Comment by chenmoneygithub** [comment]
@mattdangerw I think one thing hard to cover in a keras-io tutorial is how to create and configure the GPU cluster. There are two aspects:
- For us, we can spin up a multi-gpu cluster (maybe use GKE) so that in the future we can do testing/benchmarking there.
- For users, we can provide a keras-io tutorial using `tf.distribute` with logical devices, and also point users to how to configure multi-gpu cluster on GCP. 

I believe Reed (reedwm@) has tried our BERT model on multi-gpu cluster, we can probably use his knowledge.

**Comment by mattdangerw** [comment]
> For users, we can provide a keras-io tutorial using tf.distribute with logical devices, and also point users to how to configure multi-gpu cluster on GCP.

Yeah, let's keep this issue focused on the user documentation side of things, and not testing/benching marking.

If kaggle provides a multi-device setup, that could be a great place to demo. On keras.io faking a multi-device setup on a colab and giving GCP instructions SGTM.

**Comment by amangupta201** [comment]
@mattdangerw We can do this by using distribution strategies to run the GPUs

**Comment by chenmoneygithub** [comment]
@shivance Since data parallelism is mostly useful for large-scale training (as a comparison, users may not bother doing it for glue/mrpc), let's try doing either Masked LM training or Causal LM training. For demo purpose, we don't need to make the data volume big, for the dataset, you can find any text dataset. To go a step further, we can show these:
- Collect a text dataset and convert to `tf.data.Dataset`.
- Use `MLMMaskGenerator` to generate masked LM training set.
- Use `BertMaskedLM` to do the training.
- Use `MirroredStrategy`/`MultiworkerMirroredStrategy` + logical device to parallel the data for training.
- Show some logging for parallelism so that users know they are actually doing distributed training.

This proposal contains some extra efforts for building the MLM training pipeline, so if we want to focus more on the distributed training part, here is another proposal:
- Use GLUE/MNLI for our demo.
- Use `BertClassifier` for finetuning.
- Use `MirroredStrategy`/`MultiworkerMirroredStrategy` to distribute the data.

Please let me know which one you are more interested in!

@mattdangerw Tag Matt for comments.

**Comment by shivance** [comment]
The former one interests me, not only because it has stuff to learn in store but also because it'll demonstrate usage of our MaskedLM API, so it creates a lot of value. Let's not forget what @mattdangerw said [here](https://github.com/keras-team/keras-nlp/discussions/775#discussioncomment-5105254) , When is 0.5 release anyway? Shipping this keras-io example with the release might make it more grand, what say @chenmoneygithub ?

**Comment by mattdangerw** [comment]
> When is 0.5 release anyway?

Not totally sure, but probably a couple weeks out!

Demoing distributed MaskedLM training sounds great to me! Let's plan on having this on keras.io if we can, but also cross posting on kaggle sounds fine if of interest.

I'm not totally sure what the overall flow should be, the tricky part will be figure out how to explain a multi-accelerator setup without necessarily having it in the colab environment used to run the example.

We could either show a multi-GPU, single device setup, or a full multi-machine setup, which is documented here https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras.

An ambitious idea would be to start with a normal training example (no distribution strategy), then show `tf.distribute.MirroredStrategy` for multi-GPU, then show `tf.distribute.MultiWorkerMirroredStrategy` for multi-machine, all in the same guide. That could be a very cool progression! For the last part, we would probably just want to show the "setup code" and a few training steps, not attempt to really train.

I think it would also be very useful to give some description of how you could spin up a cluster of machines on GCP to do this type of training, @chenmoneygithub might have some insight there.

**Comment by shivance** [comment]
1. 
>     * Use `MLMMaskGenerator` to generate masked LM training set.
@chenmoneygithub Doesn't `BertMaskedLMPreprocessor do it for us out of the box ? I'm using a model from preset.

2. [Notebook Link]([Notebook ](https://www.kaggle.com/code/shivanshuman/tf-distribute) , does anyone have idea what's the source of error in last cell of notebook? I'm using two T4 GPUs on kaggle

![image](https://user-images.githubusercontent.com/51750587/228937288-2edc2f53-11ea-4355-9f12-3a9f2930b8cd.png)

**Comment by shivance** [comment]
[Update] : Fix to above error is avoid using

```
train_dist_ds = strategy.experimental_distribute_dataset(train_ds)
val_dist_ds = strategy.experimental_distribute_dataset(val_ds)````

And use `train_ds` and `val_ds` directly
